tla-io/src/main/scala/at/forsyte/apalache/io/itf/ItfToTla.scala

```diff
@@ -0,0 +1,249 @@
+package at.forsyte.apalache.io.itf
+
+import at.forsyte.apalache.io.itf.ItfToTla._
+import at.forsyte.apalache.io.json.{JsonDeserializationError, JsonRepresentation, ScalaFactory}
+import at.forsyte.apalache.io.lir.Trace
+import at.forsyte.apalache.tla.lir._
+import at.forsyte.apalache.tla.typecomp.TBuilderInstruction
+import at.forsyte.apalache.tla.types.parser.DefaultType1Parser
+import at.forsyte.apalache.tla.types.{tla, ModelValueHandler}
+
+/**
+ * Reads a sequence of States, i.e. mappings from variable names to TlaEx values, from an ITF json.
+ *
+ * @tparam T
+ *   Any class extending JsonRepresentation
+ * @author
+ *   Jure Kukovec
+ */
+class ItfToTla[T <: JsonRepresentation](
+    scalaFactory: ScalaFactory[T]) {
+
+  /**
+   * Checks that the Apalache-specific ITF requirements are met, as defined in ADR15. If yes, returns a mapping from
+   * variables to their types.
+   */
+  private[itf] def validateShapeAndGetTypes(json: T): Map[String, TlaType1] = {
+    val varTypesOpt = for {
+      meta <- json.getFieldOpt(META_FIELD)
+      varTypes <- meta.getFieldOpt(VAR_TYPES_FIELD)
+    } yield varTypes
+    val varTypesJson = varTypesOpt.getOrElse {
+      throw new JsonDeserializationError("ITF JSON does not satisfy Apalache requirements. " +
+        "Toplevel \"#meta\" field must exist and contain \"varTypes\". " +
+        "See ADR15: https://apalache.informal.systems/docs/adr/015adr-trace.html#itf-as-input-to-apalache. ")
+    }
+
+    val diffOpt = for {
+      vars <- json.getFieldOpt(VARS_FIELD)
+      varTypesKeys <- varTypesJson.allFieldsOpt
+    } yield {
+      val varsSet = scalaFactory.asSeq(vars).map { scalaFactory.asStr }.toSet
+      // Symmetric difference
+      (varsSet -- varTypesKeys) ++ (varTypesKeys -- varsSet)
+    }
+
+    val diff = diffOpt.getOrElse {
+      throw new JsonDeserializationError(s"\"$VARS_FIELD\" or \"$VAR_TYPES_FIELD\" contents are malformed.")
+    }
+    if (diff.nonEmpty)
+      throw new JsonDeserializationError(s"\"$VARS_FIELD\" and \"$VAR_TYPES_FIELD\" define different states. " +
```

Comment:
Wouldn't a JSON deserialization error would occur prior to obtaining a `ujson.Value`?. This is rather a data-validation error.

Not sure how you want to do error handling here, but I wouldn't wrap it up with JSON deserialization.

Reply:
Well. that's a matter of perspective. For me, the step "read a trace from a json file" is, in a sense, atomic, so anything that happens inside the code that does that "failed to deserialize" the json. It's also the most useful for the user, the end message should say that something is wrong with the json they've provided.